### PR TITLE
Implemented feature #3: Validate expense

### PR DESCRIPTION
**Pull request by ichbinju0**
We solve problem #3 

1. check if 'payer' is nonblank and not null
2. check if 'amount' is bigger than zero and 'amount' is not null
3. check if 'participants' size is bigger or same than 1 and if 'names' are nonblank and not null
4. check if 'currency' is nonblank and not null
5. check if 'date' is nonblank and not null
-> if values are not validate, our code throws 'IllegalArgumentException' with messages

Also, we made test cases:
All tests use a helper method 'isValid'
1. happy path test:
-  'validExpense': we tested standard 'expense' object with valid fields to ensure it passes validation
- 'testValidMultipleParticipants': we confirm that an expense with a larger list of participants correctly passes validation
2. edge case test: we both checked null and blank values seperately for every fields

### CODE REVIEW

**Comment by hayat711** [comment]
Group1 will review this PR [code implemented by group2 for issue #3 ]

File: src/main/java/itm/oss/splitter/ExpenseValidator.java [hunk 1]
```diff
@@ -1,8 +1,43 @@
...
+
 public class ExpenseValidator {
   public static void validate(Expense e) {
-    // TODO (Issue 3): implement checks (payer nonblank, amount > 0, participants nonempty, currency nonblank)
-    // Keep empty for Phase 1.
+    // TODO (Issue 3): implement checks (payer nonblank, amount > 0, participants
+    // nonempty, currency nonblank)
+
+    // check if payer is nonblank and not null
+    if (e == null || e.getPayer().isBlank()) {
```
Comment (ehdgus4173) [inline src/main/java/itm/oss/splitter/ExpenseValidator.java]: Since `e == null` is written in every conditional, it might be better to make a new exception that checks `e` is null.

File: src/main/java/itm/oss/splitter/ExpenseValidator.java [hunk 2]
```diff
@@ -1,8 +1,43 @@
...
+
+    }
+
+    // check if amount is bigger than zero and amount is not null
+    if (e == null || e.getAmount().compareTo(BigDecimal.ZERO) <= 0) {
+      throw new IllegalArgumentException("Amount should be larger than 0");
+    }
+
+    // check if currency is nonblank and not null
+    if (e == null || e.getCurrency().isBlank()) {
```
Comment (ehdgus4173) [inline src/main/java/itm/oss/splitter/ExpenseValidator.java]: The structure of three conditionals that check `payer`, `currency`, `date`  are same. Making a helper method might make code easier to read.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+import org.junit.jupiter.api.Test;
+import java.math.BigDecimal;
+import java.util.ArrayList;
+import java.util.Arrays;
+import static org.junit.jupiter.api.Assertions.*;
+
+/**
+ * This is a test class for ExpenseValidator.
+ */
+public class ExpenseValidatorTest {
```
Comment (Junseop1228) [inline src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:13]: Excellent test coverage for various edge cases like null and blank inputs. The readable test names are also a great practice!

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+     * Edge Case test
+     * Checks empty participants list
+     */
+    @Test
+    @DisplayName("Empty participants list should fail")
+    void testEmptyParticipants() {
+        Expense e = new Expense("2025-10-23", "kwon hyeok yoon", new BigDecimal("100.00"),
+                "USD", new ArrayList<>(), "IT", "PC");
+        boolean result = isValid(e); // result should be false.
+        assertFalse(result, "Participants list cannot be empty.");
```
Comment (Junseop1228) [inline src/test/java/itm/oss/splitter/ExpenseValidatorTest.java]: This helper method is clever! For more robust tests, consider using assertThrows to verify the exact exception type and message.

**Review by ehdgus4173 — COMMENTED** [review]
The code all looks great. Just suggested few changes about code duplication!

File: src/main/java/itm/oss/splitter/ExpenseValidator.java [hunk 5]
```diff
@@ -1,8 +1,43 @@
...
-    // TODO (Issue 3): implement checks (payer nonblank, amount > 0, participants nonempty, currency nonblank)
-    // Keep empty for Phase 1.
+    // TODO (Issue 3): implement checks (payer nonblank, amount > 0, participants
+    // nonempty, currency nonblank)
+
+    // check if payer is nonblank and not null
+    if (e == null || e.getPayer().isBlank()) {
+      throw new IllegalArgumentException("Payer is Empty! Please add payer");
+
+    }
```
Comment (byungchan3077) [inline src/main/java/itm/oss/splitter/ExpenseValidator.java]: This method is clearly defined! However there is some confusing point of throwing exception. When even e is equal to null, it always throws the exception that "Payer is Empty! Please add payer". It would be more readable  if separate the cases.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java [hunk 6]
```diff
@@ -0,0 +1,217 @@
...
+     */
+    @Test
+    @DisplayName("Null date should fail")
+    void testNullDate() {
+        Expense e = new Expense(null, "RM", new BigDecimal("100.00"),
+                "USD", new ArrayList<>(Arrays.asList("Jin")), "Music", "Concert");
+        boolean result = isValid(e); //result should be false
+        assertFalse(result, "Date cannot be blank.");
+    }
+
```
Comment (byungchan3077) [inline src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:243]: The edge case tests are very well designed, and how about you add a edge case, which is that the object 'expense' itself is null?

**Review by byungchan3077 — COMMENTED** [review]
Suggestions for Improving Validator Robustness and Test Cases

**Comment by jobregon1212** [comment]
Hello guys, this PR should only modify the .java files (implementation and test files). It should not touch .gitignore , README, or any other file. So please fix that.  Remember to synchronize with upstream repository and pull the latest changes from the upstream repository.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 6 above)
Comment (Gyeongyoon) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:243]: Thank you for the suggestion! I added an edge case test for when the Expense object itself is null.
스크린샷 2025-11-04 오후 7 29 55

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 5 above)
Comment (Gyeongyoon) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: Thanks for the feedback! This issue was resolved when my teammate modified this part — the null check for the Expense object is now handled separately at the top, so the cases are clearly separated.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 3 above)
Comment (so-yeon-333) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:13]: Thanks for your kind review! I am very happy to see your positive comment!

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 1 above)
Comment (ichbinju0) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: Thank you for your feed back. There was a problem that e = null is duplicate. I changed the code to check null only once in the top of the code.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 4 above)
Comment (so-yeon-333) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java]: Thanks for your feedback. I changed all the testcases to use the assertthrows. We can now perform more precise testing. I hope you like the changed version.

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 2 above)
Comment (ichbinju0) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: Thank you for your feedback. As you mentioned, the logic for validating payer, currency, and date was repeatedly applied to String types.I refactored this part by extracting it into a helper method called checkStringIsBlank to remove redundancy and improve readability.

The validation logic for Amount and Participants has different types and validation methods, and since they are only used once within the validate method, I kept them inline for better readability instead of creating separate helper methods.
Please review the updated code!

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 1 above)
Comment (ehdgus4173) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: > Thank you for your feed back. There was a problem that e = null is duplicate. I changed the code to check null only once in the top of the code.

I checked your changes. Thank you for accepting my comment, LGTM!

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 2 above)
Comment (ehdgus4173) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: > Thank you for your feedback. As you mentioned, the logic for validating payer, currency, and date was repeatedly applied to String types.I refactored this part by extracting it into a helper method called checkStringIsBlank to remove redundancy and improve readability.
> 
> The validation logic for Amount and Participants has different types and validation methods, and since they are only used once within the validate method, I kept them inline for better readability instead of creating separate helper methods. Please review the updated code!

Thank you for your agreement!  Your changes look very valid. 👍

**Review by ehdgus4173 — COMMENTED** [review]
Checked your changes. Thank you for your agreement! LGTM👍

File: src/main/java/itm/oss/splitter/ExpenseValidator.java (hunk 5 above)
Comment (byungchan3077) [reply src/main/java/itm/oss/splitter/ExpenseValidator.java]: Okay, I think the code became more clear. Thank you for your comments.

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 6 above)
Comment (byungchan3077) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:243]: LGTM. The test code looks perfect. Thank you for considering my suggestion!

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 3 above)
Comment (Junseop1228) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java:13]: I checked your improvement even though it was already really good. Well done !

File: src/test/java/itm/oss/splitter/ExpenseValidatorTest.java (hunk 4 above)
Comment (Junseop1228) [reply src/test/java/itm/oss/splitter/ExpenseValidatorTest.java]: The test become better. Looks very good to me.

**Review by Junseop1228 — APPROVED** [review]
Final review on changes.

**Review by byungchan3077 — APPROVED** [review]
I saw all the modifications, and It looks better to me.

**Review by ehdgus4173 — APPROVED** [review]
Approved review. LGTM!
